### PR TITLE
Feature/LJI-18-plans-filters

### PR DESCRIPTION
**Pull request by NormantasN**
This PR introduces the following features for the Plans:

1. Filtering plans allow filtering by plan name, status, plan type and also multiple filters (e.g filter by name and status at the same time)

2. Sorting plans allow sorting by default (by name), by type or status, ascending or descending

3. Pagination - 10 entities on a page by default

4. New endpoint:
- /v1/plans/filtered for handling filtered, sorted, and paginated requests
- Postman requests
- Unit and integration tests

### CODE REVIEW

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java [hunk 1]
```diff
@@ -1,116 +1,144 @@
...
+  @Test
+  void testGetAllPlans() {
+    List<PlanDTO> plansList = easyRandom.objects(PlanDTO.class, 3).toList();
+    when(planService.getAllPlans()).thenReturn(plansList);
+
+    List<PlanDTO> result = planService.getAllPlans();
 
-    @Mock
-    private PlanService planService;
+    assert result != null;
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:54]:
```suggestion
    assertNotNull(result);
```

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java [hunk 2]
```diff
@@ -1,116 +1,144 @@
...
+  private final EasyRandom easyRandom = new EasyRandom();
+
+  @Mock
+  private PlanService planService;
+
+  @Test
+  void testAddPlan() {
+    PlanDTO newPlanDTO = easyRandom.nextObject(PlanDTO.class);
+
+    when(planService.addPlan(any(PlanDTO.class))).thenReturn(newPlanDTO);
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:39]: This test class is created for test service, but you mocked it. 
Please reformat your test class fully to test in the right way your service

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java [hunk 3]
```diff
@@ -129,5 +133,21 @@ void testDeactivatePlan() throws Exception {
     verify(planService).deactivatePlan(id);
   }
 
+  @Test
+  @WithMockUser(username = "admin")
+  public void testGetFilteredSortedPlans() throws Exception {
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java:138]: Please reformat fully your test class regarding this comment - https://github.com/Java-POT-Summer-24/pot-java-summer-24/pull/24#discussion_r1665361282

**Review by SavaLX — CHANGES_REQUESTED** [review]

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java (hunk 1 above)
Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:54]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java (hunk 2 above)
Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:39]: How should I test it without using Mockito.when?

Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:39]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java (hunk 3 above)
Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java:175]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java [hunk 4]
```diff
@@ -1,116 +1,131 @@
...
+@Transactional
 class PlanServiceTest {
 
-    private final EasyRandom easyRandom = new EasyRandom();
+  private final EasyRandom easyRandom = new EasyRandom();
 
-    @Mock
-    private PlanService planService;
+  @Autowired
+  private PlanService planService;
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java:38]: As discussed in the stand-up, mock the rep layer and unit test the service layer(business logic)

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java (hunk 2 above)
Comment (SavaLX) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java]: As discussed in the stand-up, mock the rep layer and unit test the service layer(business logic). Use @ExtendWith(MockitoExtension.class)

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java (hunk 3 above)
Comment (SavaLX) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java:175]: See my example of an Integration test with a mocking rep layer  in the chat

Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java:137]: Fixed

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java (hunk 2 above)
Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanServiceTest.java]: Fixed

File: src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java [hunk 5]
```diff
@@ -2,35 +2,41 @@
...
-import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors;
-import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.context.TestPropertySource;
 import org.springframework.test.web.servlet.MockMvc;
 
-@WebMvcTest(PlanController.class)
-@ExtendWith(SpringExtension.class)
+@WebMvcTest({PlanController.class, PlanService.class})
+@TestPropertySource(properties = "spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration")
 public class PlanControllerIntegrationTest {
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java:40]: Redundant 'Controller' word in the name

Comment (NormantasN) [reply src/test/java/com/coherentsolutions/pot/insurance/plan/PlanControllerIntegrationTest.java]: Done

**Review by SavaLX — APPROVED** [review]

**Review by anddevdev — APPROVED** [review]

**Pull request merged**